// chix-flow/lib/context/blessedProvider.js
var blessed = require('blessed'), EventEmitter = require('events').EventEmitter;

var BlessedProvider = (function () {
    function BlessedProvider() {
        this.questions = [];
        this.questioning = false;
        this.screen = undefined;
    }
    BlessedProvider.prototype.addContext = function (node, defaultContext) {
        this.screen = new blessed.screen();

        node.inputTimeout = 120000;

        if (typeof defaultContext !== "undefined") {
            var question = this.askForContext(node, defaultContext);

            if (!this.questioning) {
                question(node, defaultContext);
            } else {
                this.questions.push(question);
            }
        }
    };

    BlessedProvider.prototype.askForContext = function (node, defaultContext) {
        var self = this;

        return function () {
            self.questioning = true;

            var name = Object.keys(defaultContext);

            widgets.InputBox('Enter ' + defaultContext[name] + ':', {
                pos: 'center',
                style: {
                    colors: {
                        bg: 'blue',
                        input: {
                            fg: 'red',
                            bg: 'black'
                        }
                    }
                }
            }, function (input) {
                if (!input)
                    input = '';

                self.win.centertext(0, 'You entered: ' + input);

                if (typeof defaultContext[name] === "object") {
                    input = JSON.parse(input);
                }

                node.addContextProperty(name, input);

                self.win.refresh();

                setTimeout(function () {
                    self.win.close();

                    if (self.questions.length) {
                        var question = self.questions.pop();
                        question();
                    } else {
                        self.questioning = true;
                    }
                }, 1000);
            });
        };
    };
    return BlessedProvider;
})();

NcursesProvider.prototype.__proto__ = EventEmitter.prototype;

module.exports = NcursesProvider;
